src/vivarium_census_prl_synth_pop/tools/make_results.py

```diff
@@ -548,3 +554,58 @@ def subset_results_by_state(processed_results_dir: str, state: str) -> None:
             )
             write_to_disk(state_data, output_file_path)
         logger.info(f"Finished writing {observer} files for {state_name}.")
+
+
+def write_shard_metadata(
+    observer: str, obs_data: pd.DataFrame, obs_dir: Path, seed: str
+) -> None:
+    # Writes metadata for each shard of data. This will be proportions available to noise
+    # for specific columns.
+
+    def _get_metadata_values(df: pd.DataFrame, location: str):
+        metadata_df = pd.DataFrame(index=[location])
+        metadata_df["number_of_rows"] = len(df)
+        for column in METADATA_COLUMNS:
+            if column not in df.columns:
+                continue
+            elif f"copy_{column}" in COPY_HOUSEHOLD_MEMBER_COLS.values():
+                # Proportion based on missingness from available household members that can
+                # have a copy value
+                print(
+                    f"{column} in output. Calculating metadata proportion for copy_{column}..."
+                )
+                metadata_df[f"{column}_copy_number_of_rows"] = (
+                    df[f"copy_{column}"].notna().sum()
+                )
+            elif column == "first_name":
+                # Proportion of first names that have an associated nickname to be noised
+                # Everyone should have a first name so there is no missingness
+                nicknames = load_nicknames_data()
+                metadata_df[f"{column}_nicknames_number_of_rows"] = (
+                    df[column].isin(nicknames.index).sum()
+                )
+            # TODO: Add guaridan based duplication
+            else:
+                raise ValueError(f"Column '{column}' not supported for metadata recording.")
+        return metadata_df
+
+    # Special case SSA dataset since that does not have a state column
+    if observer == metadata.DatasetNames.SSA:
+        shard_metadata = _get_metadata_values(obs_data, "USA")
+    else:
+        groupby_col = "state" if "state" in obs_data.columns else "mailing_address_state"
+        location_groups = obs_data.groupby(groupby_col)
```

Comment:
Won't this exclude a state if it is not present in a given shard? You will need to make sure you account for this when you compute the overall values.

Reply:
Correct. When I compute overall values wouldn't that just mean that the shard missing a state would be 0 for that state and not affect the overall values?

Reply:
Yes, the counts will be 0